src/hip_module.cpp

```diff
@@ -313,6 +325,50 @@ hipError_t hipHccModuleLaunchKernel(hipFunction_t f, uint32_t globalWorkSizeX,
         localWorkSizeZ, sharedMemBytes, hStream, kernelParams, extra, startEvent, stopEvent, 0));
 }
 
+hipError_t hipExtLaunchMultiKernelMultiDevice(hipLaunchParams* launchParamsList,
+                                              int  numDevices, unsigned int  flags) {
+
+    hipError_t result;
+    int nGPUs, currentDevice;
+
+    hipGetDevice(&currentDevice);
+    hipGetDeviceCount(&nGPUs);
+
+    if ((numDevices > nGPUs) || (launchParamsList == nullptr)) {
+        return hipErrorInvalidValue;
+    }
+
+    for (int i = 0; i < numDevices; ++i) {
+        hipSetDevice(i);
+        hipDeviceSynchronize();
+    }
+
+    for (int i = 0; i < numDevices; ++i) {
+        hipSetDevice(i);
+        const hipLaunchParams& lp = launchParamsList[i];
+        hipFunction_t kd = hip_impl::get_program_state().kernel_descriptor(reinterpret_cast<std::uintptr_t>(lp.func),
+                hip_impl::target_agent(lp.stream));
+        if (kd == nullptr) {
+            return hipErrorInvalidValue;
+        }
+        hip_impl::kernargs_size_align kargs = hip_impl::get_program_state().get_kernargs_size_align(
+                reinterpret_cast<std::uintptr_t>(lp.func));
+        kd->_kernarg_layout = *reinterpret_cast<const std::vector<std::pair<std::size_t, std::size_t>>*>(
+                kargs.getHandle());
+
+        result = ihipModuleLaunchKernel(kd,
```

Comment:
I am not sure how we are ensuring the kernels are enqueued atomically. My expectation was that we lock all the streams first, then enqueue all the kernels followed by unlocking all the streams in the end i.e. there would be 3 for loops. Maybe I am missing something here?

Reply:
I made changes to lock all the streams before dispatching kernels in the main loop (i.e., calling ihipModuleLaunchKernel for each device) and unlock them after that to make sure that kernels are en-queued atomically on all devices.